### PR TITLE
Keyword context: work with arbitrary user input

### PR DESCRIPTION
**Pull request by camdencheek**
We were previously concatenating the user query to our search query. The problem with this is the search query has syntax rules, and the user query can be anything. To fix this, we pass the quoted user query via the `content:` field, which can handle special characters or keywords.

## Test plan

Added a unit test + manual testing that it fixes the regex error

### CODE REVIEW

File: enterprise/cmd/frontend/internal/context/resolvers/context.go [hunk 1]
```diff
@@ -39,7 +41,13 @@ func (r *Resolver) GetCodyContext(ctx context.Context, args graphqlbackend.GetCo
...
 	repoNameIDs := make([]types.RepoIDName, len(repoIDs))
 	for i, repoID := range repoIDs {
-		repoNameIDs[i] = types.RepoIDName{ID: repoID, Name: repos[repoID].Name}
+		repo, ok := repos[repoID]
+		if !ok {
+			// GetReposSetByIDs does not error if a repo could not be found.
+			return nil, errors.Newf("could not find repo with id %d", int32(repoID))
+		}
+
+		repoNameIDs[i] = types.RepoIDName{ID: repoID, Name: repo.Name}
```
Comment (camdencheek) [inline enterprise/cmd/frontend/internal/context/resolvers/context.go:50]: Unrelated issue I came across while debugging: `GetReposSetByIDs` does not error if the ids do not exist, so we can't depend on the repo set being complete.

File: internal/search/keyword/query_transformer.go [hunk 2]
```diff
@@ -53,7 +50,10 @@ func nodeToPatternsAndParameters(rootNode query.Node) ([]string, []query.Paramet
 					patterns = append(patterns, concatNodeToPatterns(op)...)
 				}
 			case query.Parameter:
-				if op.Field != query.FieldCount && op.Field != query.FieldCase && op.Field != query.FieldType {
+				if op.Field == query.FieldContent {
+					// Split any content field into a set of patterns ignoring any punctuation and special characters
+					patterns = append(patterns, keywordRegex.FindAllString(op.Value, 16)...)
+				} else if op.Field != query.FieldCount && op.Field != query.FieldCase && op.Field != query.FieldType {
```
Comment (camdencheek) [inline internal/search/keyword/query_transformer.go:56]: Now that we're passing user queries via `content:`, keyword search needs to respect that. This pulls all the `content:` fields out of the parameters, splits them into simple words, and adds them to our list of "patterns." Definitely slightly hacky, but should work more consistently

Comment (jtibshirani) [reply internal/search/keyword/query_transformer.go:56]: This makes sense to me! I'm not sure this matches the existing tokenization for keyword search though:
* Split on whitespace
* Preserve special characters inside a token 

So for example "Is crewjam/saml used anywhere" would be converted to [is, crewjam/saml, used, anywhere].

File: enterprise/internal/codycontext/context.go [hunk 3]
```diff
@@ -190,8 +192,8 @@ func (c *CodyContextClient) getKeywordContext(ctx context.Context, args GetConte
 		regexEscapedRepoNames[i] = regexp.QuoteMeta(string(repo.Name))
 	}
 
-	textQuery := "repo:^" + query.UnionRegExps(regexEscapedRepoNames) + "$ " + textFileFilter + " " + args.Query
-	codeQuery := "repo:^" + query.UnionRegExps(regexEscapedRepoNames) + "$ -" + textFileFilter + " " + args.Query
+	textQuery := fmt.Sprintf(`repo:^%s$ %s content:%s`, query.UnionRegExps(regexEscapedRepoNames), textFileFilter, strconv.Quote(args.Query))
+	codeQuery := fmt.Sprintf(`repo:^%s$ -%s content:%s`, query.UnionRegExps(regexEscapedRepoNames), textFileFilter, strconv.Quote(args.Query))
```
[the remark this answers is not in the document]
Comment (thenamankumar) [reply enterprise/internal/codycontext/context.go:196]: Have you tested if the keyword search extracts the keyword correctly from the `content:""` having code in the text?

[the remark this answers is not in the document]
Comment (thenamankumar) [reply enterprise/internal/codycontext/context.go:196]: Cool, saw the corresponding changes to support this below.

**Review by thenamankumar — APPROVED** [review]

File: internal/search/keyword/query_transformer.go (hunk 2 above)
Comment (camdencheek) [reply internal/search/keyword/query_transformer.go:56]: Do you mind pointing me to where this happens? AFAICT, we just rely on search query tokenization to split into patterns

Comment (camdencheek) [reply internal/search/keyword/query_transformer.go:56]: Second question: does that tokenization strategy make sense for code? e.g. `myFunction(abc)` wouldn't provide great keywords if we didn't split it

Comment (camdencheek) [reply internal/search/keyword/query_transformer.go]: Sounds good, updated the PR to just split the input on whitespace

File: internal/search/keyword/query_transformer.go [hunk 4]
```diff
@@ -53,7 +47,10 @@ func nodeToPatternsAndParameters(rootNode query.Node) ([]string, []query.Paramet
 					patterns = append(patterns, concatNodeToPatterns(op)...)
 				}
 			case query.Parameter:
-				if op.Field != query.FieldCount && op.Field != query.FieldCase && op.Field != query.FieldType {
+				if op.Field == query.FieldContent {
+					// Split any content field into a set of patterns ignoring any punctuation and special characters
```
Comment (jtibshirani) [inline internal/search/keyword/query_transformer.go:51]: Should this just say "Split any content field on whitespace"?

**Review by jtibshirani — APPROVED** [review]

**Pull request merged**